File: packages/android/engine/src/main/kotlin/me/sospedra/blasphem/Judge.kt
package me.sospedra.blasphem

import android.content.Context
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * A judge built once and called on every keystroke.
 *
 * [create] reads the packs from the app assets, so call it off the main
 * thread. [judge] is synchronous and safe from several threads. [close]
 * releases the engine.
 */
class Judge private constructor(handle: Long) : AutoCloseable {
    /** The loaded locales, in registry order. */
    val locales: List<String> = Native.engineLocales(handle).toList()

    private var engine: Long = handle
    private val lock = ReentrantReadWriteLock()

    /** Scores one message. Never throws while the judge is open; throws [BlasphemException.Code.CLOSED] after [close]. */
    fun judge(text: String): Judgement = lock.read {
        if (engine == 0L) throw BlasphemException(BlasphemException.Code.CLOSED, "the judge was closed")
        Native.engineJudge(engine, text)
    }

    /** Releases the engine. Later [judge] calls throw. Calling it twice is harmless. */
    override fun close() = lock.write {
        if (engine == 0L) return@write
        Native.engineFree(engine)
        engine = 0L
    }

    companion object {
        /** Reads the Gradle configuration and obtains the complete selected release. */
        @JvmStatic
        suspend fun create(context: Context): Judge {
            val (config, sources) = withContext(Dispatchers.IO) {
                val config = try { BundleConfiguration.read(context) }
                catch (failure: Exception) {
                    throw BlasphemException(BlasphemException.Code.FETCH_FAILED, "bundle.json: ${failure.message}")
                }
                val read = try { config.sources(context) }
                catch (failure: Exception) {
                    throw BlasphemException(BlasphemException.Code.FETCH_FAILED, "language data: ${failure.message}")
                }
                val sources = config.locales.map { Source(it, read, config.detectLanguage) }
                config to sources
            }
            return suspendCancellableCoroutine { continuation ->
                if (continuation.isActive) {
                    val judge = Judge(buildEngine(sources, JudgeOptions(config.locales, config.detectLanguage)))
                    continuation.resume(judge) { _, value, _ -> value.close() }
                }
            }
        }

        /**
         * Loads `options.locales` from `context.assets`, or from `options.packsDirectory`
         * when set, and builds the engine. Blocks on file reads.
         */
        @JvmStatic
        fun create(context: Context, options: JudgeOptions): Judge {
            val codes = normalizeLocales(options.locales)
            val read = options.packsDirectory?.let(::directoryReader) ?: assetReader(context.assets)
            val sources = codes.map { code -> Source(code, read, options.detectLanguage) }
            return Judge(buildEngine(sources, options))
        }
    }
}

private class Source(val code: String, read: ReadArtifact, detectLanguage: Boolean) {
    val pack: ByteArray = read(code, ArtifactKind.PACK)
    val detect: ByteArray? = if (detectLanguage) read(code, ArtifactKind.DETECT) else null
}

private fun buildEngine(sources: List<Source>, options: JudgeOptions): Long {
    val builder = Native.builderNew(options.detectLanguage, options.grawlix)
    try {
        for (source in sources) Native.builderAdd(builder, source.code, source.pack, source.detect)
        return Native.builderBuild(builder)
    } catch (failure: RuntimeException) {
        Native.builderFree(builder)
        throw BlasphemException.fromEngine(failure.message ?: "the native builder failed without a message")
    }
}

File: packages/apple/Sources/Blasphem/Judge.swift
import BlasphemFFI
import Foundation

/// A judge built once and called on every keystroke.
///
/// Construction reads the packs from disk, so call it off the main thread.
/// `judge(_:)` is synchronous and safe from several threads. `close()`
/// releases the engine; `deinit` closes too.
public final class Judge: @unchecked Sendable {
    /// The loaded locales, in registry order.
    public let locales: [String]

    private var engine: OpaquePointer?
    private let lock = ReadWriteLock()

    /// Loads `locales` and builds the engine.
    ///
    /// - Parameters:
    ///   - locales: lowercase codes such as `["en", "es"]`, `id` (Indonesian), or `ms` (Malay). Empty throws.
    ///   - detectLanguage: route by detected language. Every locale then needs its detection slice.
    ///   - grawlix: populate `Judgement.grawlix` for unsafe verdicts.
    ///   - packsDirectory: read `<code>.pack` and `<code>.detect` from this folder instead of the app bundle.
    public init(
        locales requested: [String],
        detectLanguage: Bool = true,
        grawlix: Bool = false,
        packsDirectory: URL? = nil
    ) throws {
        let codes = try normalizeLocales(requested)
        let sources = try codes.map { code in
            try Source(code: code, detectLanguage: detectLanguage, packsDirectory: packsDirectory)
        }
        let engine = try buildEngine(sources, detectLanguage: detectLanguage, grawlix: grawlix)
        self.engine = engine
        self.locales = loadedLocales(engine)
    }

    deinit {
        close()
    }

    /// Scores one message. Never fails while the judge is open; throws `.closed` after `close()`.
    public func judge(_ text: String) throws -> Judgement {
        try lock.read {
            guard let engine else {
                throw BlasphemError(code: .closed, message: "the judge was closed")
            }
            let verdict = blasphem_engine_judge(engine, text)
            defer { blasphem_judgement_free(verdict) }
            return Judgement(
                safe: verdict.safe,
                score: verdict.score,
                locale: verdict.locale.map { String(cString: $0) },
                grawlix: verdict.grawlix.map { String(cString: $0) }
            )
        }
    }

    /// Releases the engine. Later `judge(_:)` calls throw `.closed`. Calling it twice is harmless.
    public func close() {
        lock.write {
            guard let engine else { return }
            blasphem_engine_free(engine)
            self.engine = nil
        }
    }
}

private struct Source {
    let code: String
    let pack: Data
    let detect: Data?

    init(code: String, detectLanguage: Bool, packsDirectory: URL?) throws {
        self.code = code
        pack = try readArtifact(code: code, kind: .pack, packsDirectory: packsDirectory)
        detect = try detectLanguage ? readArtifact(code: code, kind: .detect, packsDirectory: packsDirectory) : nil
    }
}

private func buildEngine(_ sources: [Source], detectLanguage: Bool, grawlix: Bool) throws -> OpaquePointer {
    guard let builder = blasphem_builder_new(detectLanguage, grawlix) else {
        throw BlasphemError(code: .packInvalid, message: "the native builder could not be created")
    }
    do {
        for source in sources {
            try add(source, to: builder)
        }
        guard let engine = blasphem_builder_build(builder) else {
            throw builderError(builder)
        }
        return engine
    } catch {
        blasphem_builder_free(builder)
        throw error
    }
}

private func add(_ source: Source, to builder: OpaquePointer) throws {
    let status = withBytes(source.pack) { pack, packCount in
        withBytes(source.detect) { detect, detectCount in
            blasphem_builder_add(builder, source.code, pack, packCount, nil, detect, detectCount, nil)
        }
    }
    guard status == 0 else {
        throw builderError(builder)
    }
}

private func withBytes<Result>(_ data: Data?, _ body: (UnsafePointer<UInt8>?, Int) -> Result) -> Result {
    guard let data else { return body(nil, 0) }
    return data.withUnsafeBytes { raw in
        body(raw.bindMemory(to: UInt8.self).baseAddress, raw.count)
    }
}

private func builderError(_ builder: OpaquePointer) -> BlasphemError {
    guard let text = blasphem_builder_error(builder) else {
        return BlasphemError(code: .packInvalid, message: "the native builder failed without a message")
    }
    return BlasphemError.fromEngine(String(cString: text))
}

private func loadedLocales(_ engine: OpaquePointer) -> [String] {
    (0..<blasphem_engine_locale_count(engine)).compactMap { index in
        guard let text = blasphem_engine_locale(engine, index) else { return nil }
        defer { blasphem_text_free(text) }
        return String(cString: text)
    }
}
